addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java

```diff
@@ -498,4 +500,55 @@ public void handleConfigurationUpdate(Map<String, Object> configurationParameter
             logger.error("Error setting thing properties: {}", ex.getMessage(), ex);
         }
     }
+
+    @Override
+    public synchronized void handleRemoval() {
+        final Bridge bridge = getBridge();
+        if (bridge != null && bridge.getConfiguration().as(HomematicConfig.class).isUnpairOnDeletion()) {
+            deviceDeletionPending = true;
+            ((HomematicBridgeHandler) bridge.getHandler()).deleteFromGateway(UidUtils.getHomematicAddress(thing), false,
+                    false, true);
+        } else {
+            super.handleRemoval();
+        }
+    }
+
+    /**
+     * Called by the bridgeHandler when this device has been removed from the gateway.
+     */
+    public synchronized void deviceRemoved() {
+        deviceDeletionPending = false;
+        if (getThing().getStatus() == ThingStatus.REMOVING) {
+            // thing removal was initiated on ESH side
+            updateStatus(ThingStatus.REMOVED);
+        } else {
+            // device removal was initiated on homematic side, thing is not removed
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.GONE);
+        }
+    }
+
+    /**
+     * Called by the bridgeHandler when the device for this thing has been added to the gateway.
+     * This is used to reconnect a device that was previously unpaired.
+     * 
+     * @param device The device that has been added to the gateway
+     */
+    public void deviceLoaded(HmDevice device) {
+        try {
+            updateStatus(device);
+        } catch (BridgeHandlerNotAvailableException ex) {
+            // ignore
+        } catch (IOException ex) {
+            logger.error("{}", ex.getMessage(), ex);
```

Comment:
I actually hoped this one also to include a context, of logging.

Also an IOException is something that is always to be expected so normally it should not be reported on the error level.

Reply:
I remember @kaikreuzer to be pretty strict about these log levels recently, so it would be really appreciated it you could improved it. I even doubt whether warnings is suitable, maybe @kaikreuzer can jump in to make sure you do it as would be the best practice